src/main/java/org/broadinstitute/hellbender/engine/spark/ShuffleJoinReadsWithRefBases.java

```diff
@@ -80,4 +81,34 @@
             return out;
         });
     }
-}
\ No newline at end of file
+
+    public static <T> JavaPairRDD<GATKRead, Tuple2<T, ReferenceBases>> addBases(final ReferenceDataflowSource referenceDataflowSource,
+                                                                                final JavaPairRDD<GATKRead, T> keyedByRead) {
+        SerializableFunction<GATKRead, SimpleInterval> windowFunction = referenceDataflowSource.getReferenceWindowFunction();
+
+        JavaPairRDD<ReferenceShard, Tuple2<GATKRead, T>> shardRead = keyedByRead.mapToPair(pair -> {
+            ReferenceShard shard = ReferenceShard.getShardNumberFromInterval(windowFunction.apply(pair._1()));
+            return new Tuple2<>(shard, pair);
+        });
+
+        JavaPairRDD<ReferenceShard, Iterable<Tuple2<GATKRead, T>>> shardiRead = shardRead.groupByKey();
+
+        return shardiRead.flatMapToPair(in -> {
+            List<Tuple2<GATKRead, Tuple2<T, ReferenceBases>>> out = Lists.newArrayList();
```

Comment:
Advise against using `Arrays.asList()` here, as we can't currently serialize the return type of that method (as we couldn't get either `UnmodifiableCollectionsSerializer` or `ArraysAsListSerializer` working for it).